holoviews/core/data.py

```diff
@@ -1308,9 +1334,288 @@ def aggregate(cls, columns, kdims, function, **kwargs):
+            sampled, int_inds = [], []
+            for d, ind in zip(dimensions, sample):
+                cdata = columns.data[d]
+                mask = cls.coord_mask(columns, cdata, ind)
+                inds = np.arange(len(cdata)) if mask is None else np.argwhere(mask)
+                int_inds.append(inds)
+                sampled.append(cdata[mask])
+            for d, arr in zip(dimensions, np.meshgrid(*sampled)):
+                data[d].append(arr)
+            for vdim, array in zip(columns.vdims, arrays):
+                flat_index = np.ravel_multi_index(tuple(int_inds), array.shape)
+                data[vdim.name].append(array.flat[flat_index])
+        concatenated = {d: np.concatenate(arrays).flatten() for d, arrays in data.items()}
+        return concatenated
+
+
+    @classmethod
+    def aggregate(cls, columns, kdims, function, **kwargs):
+        kdims = [kd.name if isinstance(kd, Dimension) else kd for kd in kdims]
+        data = {kdim: columns.data[kdim] for kdim in kdims}
+        axes = tuple(columns.get_dimension_index(kdim) for kdim in columns.kdims
+                    if kdim not in kdims)
+        for vdim in columns.vdims:
+            data[vdim.name] = np.atleast_1d(function(columns.data[vdim.name],
+                                                     axis=axes, **kwargs))
+
+        return data
+
+
+    @classmethod
+    def reindex(cls, columns, kdims, vdims):
+        if len(kdims) != columns.ndims:
+            raise ValueError('Cannot drop dimension varying dimension.')
+        if (any(kd for kd in kdims if kd not in columns.kdims) or
+            any(vd for vd in vdims if vd not in columns.vdims)):
+            return columns.clone(columns.columns()).reindex(kdims, vdims)
+        data = dict(columns.data)
+
+        dropped_vdims = ([vdim for vdim in columns.vdims
+                          if vdim not in vdims] if vdims else [])
+        for vdim in dropped_vdims:
+            del data[vdim.name]
+
+        if kdims != columns.kdims:
+            axes = [columns.kdims.index(d) for d in kdims]
+            for vdim in vdims:
+                data[vdim.name] = np.transpose(data[vdim.name], axes)
+        return data
+
+
+    @classmethod
+    def add_dimension(cls, columns, dimension, dim_pos, values, vdim):
+        raise NotImplementedError
+
+
+    @classmethod
+    def add_dimension(cls, columns, dimension, dim_pos, values, vdim):
+        if not vdim:
+            raise Exception("Cannot add key dimension to a dense representation.")
+        dim = dimension.name if isinstance(dimension, Dimension) else dimension
+        return dict(columns.data, **{dim: values})
+
+
+    @classmethod
+    def sort(cls, columns, by=[]):
+        if not by or by in [columns.kdims, columns.dimensions()]:
+            return columns.data
+        else:
+            raise Exception('Dense representation cannot be sorted, either instantiate '
+                            'in the desired order or use a sparse format.')
```

Comment:
It isn't clear what you mean by sparse format. I think 'expanded format' might be clearer as they are still dense arrays but you don't need to worry about the cartesian product...